ipywidgets/widgets/trait_types/__init__.py

```diff
@@ -0,0 +1,185 @@
+    
+    Also, we default to a trivial instance, even if args and kwargs is not specified."""
+
+    def validate(self, obj, value):
+        if isinstance(value, dict):
+            return super(InstanceDict, self).validate(obj, self.klass(**value))
+        else:
+            return super(InstanceDict, self).validate(obj, value)
+
+    def make_dynamic_default(self):
+        return self.klass(*(self.default_args or ()),
+                          **(self.default_kwargs or {}))
+
+
+class EDict(eventful.Eventful, traitlets.Dict):
+
+    event_map = {
+        'setitem': ('__setitem__', 'setdefault'),
+        'delitem': ('__delitem__', 'pop'),
+        'update': 'update',
+        'clear': 'clear',
+    }
+    type_name = 'edict'
+
+    def _before_update(self, inst, call):
+        new = call.args[0]
+        new.update(call.kwargs)
+        call_list = []
+        return self.redirect(
+            None, 'setitem', inst,
+            args=new.items())
+
+    def _before_clear(self, inst, call):
+        return self.redirect(
+            None, 'delitem', inst,
+            args=inst.keys())
+
+    @staticmethod
+    def before_setitem(inst, call):
+        """Expect call.args[0] = key"""
+        key, old = call.args[0], inst.get(call.args[0], traitlets.Undefined)
+        def after_setitem(returned):
+            new = inst.get(key, traitlets.Undefined)
+            if not equivalent(old, new):
+                return traitlets.Bunch(key=key, old=old, new=new)
+        return after_setitem
+
+    _before_setitem = before_setitem
+    _before_delitem = before_setitem
+
+
+class EList(eventful.Eventful, traitlets.List):
+
+    event_map = {
+        'append': 'append',
+        'extend': 'extend',
+        'setitem': '__setitem__',
+        'reverse': 'reverse',
+        'sort': 'sort',
+    }
+    type_name = 'elist'
+
+    def before_setitem(self, inst, index):
+        try:
+            old = inst[index]
+        except:
+            old = traitlets.Undefined
+        def after_setitem(returned):
+            try:
+                new = inst[index]
+            except:
+                new = traitlets.Undefined
+            if not equivalent(old, new):
+                return traitlets.Bunch(index=index, old=old, new=new)
+        return after_setitem
+
+    def _before_setitem(self, inst, call):
+        return self.before_setitem(inst, call.args[0])
+
+    def rearrangement(self, origin, inst):
+        return self.redirect(origin, 'setitem', inst,
+            args=[(i,) for i in range(len(inst))])
+
+    def _before_reverse(self, inst, call):
+        return self.rearrangement(inst, 'reverse')
+
+    def _before_sort(self, inst, call):
+        return self.rearrangement(inst, 'sort')
+
+    def _before_extend(self, inst, call):
+        size = len(call.args[0])
+        return self.redirect('append', 'setitem', inst,
+            args=[(i+len(inst),) for i in range(size)])
```

Comment:
I'm going to have a `redirect` decorators rather than the method:

```python
@cause("setitem")
def _before_extend(self, value, call, setitem):
    for i in range(len(call.args[0])):
        setitem(i + len(value))
```

Hopefully this is more intuitive than what's currently implemented?